Adding check for cyclic configuration

HenningNT opened the pull request:
Not sure what happened to the previous PR, it got canceled automatically when I did the repository undo...
Anyway, here's a new fresh PR, which changes one thing only.
I added two tests to verify the the check throws ArgumentException, the reviewer might want to double check the test to verify that the created state machine is actually a cyclic configuration :-D

nblumhardt commented on src/Stateless/StateConfiguration.cs:
```diff
@@ -508,6 +508,26 @@ public StateConfiguration OnExit(Action<Transition> exitAction, string exitActio
+                var State = _representation.UnderlyingState;
+
+                // Check for accidental identical cyclic configuration 
+                if (State.Equals(superstate)) { throw new ArgumentException($"Configuring {State} as a substate of {superstate} creates an illegal cyclic configuration."); }
```
Nit: apart from argument null checks, `if` statements in the repo are mostly over multiple lines. Braces optional, but if they're present they should be on a line each.

nblumhardt commented on src/Stateless/StateConfiguration.cs:
```diff
@@ -508,6 +508,26 @@ public StateConfiguration OnExit(Action<Transition> exitAction, string exitActio
+                if (State.Equals(superstate)) { throw new ArgumentException($"Configuring {State} as a substate of {superstate} creates an illegal cyclic configuration."); }
+
+                // Check for accidental identical nested cyclic configuration 
+                List<TState> superstates = new List<TState>();
```
Nit: `var superstates = new List<TState>{ State };`

nblumhardt replied:
Thinking through it a little further, could be:

```csharp
var superstates = new HashSet<TState>{ State };
```

The check for membership could then be moved inside the `while` loop as below.

nblumhardt commented on src/Stateless/StateConfiguration.cs:
```diff
@@ -508,6 +508,26 @@ public StateConfiguration OnExit(Action<Transition> exitAction, string exitActio
+                var activeRepresentation = _lookup(superstate);
+                while (activeRepresentation.Superstate != null)
+                {
+                    superstates.Add(activeRepresentation.Superstate.UnderlyingState);
```
Using `HashSet` and moving the check here we can get away from the `GroupBy` below.

```csharp
if (superstates.Contains(activeRepresentation.Superstate.UnderlyingState))
    throw new ArgumentException($"Configuring {State} as a substate of {superstate} creates an illegal nested cyclic configuration.");
```

nblumhardt commented on test/Stateless.Tests/StateMachineFixture.cs:
```diff
@@ -359,5 +359,20 @@ public void TheOnTransitionEventFiresBeforeTheOnEntryEvent()
             }
         }
+        [Test]
+        public void CyclicConfigurationDetected()
```
`DirectCyclicConfigurationDetected()`?

nblumhardt commented on test/Stateless.Tests/StateMachineFixture.cs:
```diff
@@ -359,5 +359,20 @@ public void TheOnTransitionEventFiresBeforeTheOnEntryEvent()
+        {
+            var sm = new StateMachine<State, Trigger>(State.A);
+
+            Assert.Throws(typeof(ArgumentException),  delegate { sm.Configure(State.A).SubstateOf(State.A); });
```
Is the old-style `delegate` needed?

```csharp
Assert.Throws(typeof(ArgumentException),  () => { sm.Configure(State.A).SubstateOf(State.A); });
```

nblumhardt commented on test/Stateless.Tests/StateMachineFixture.cs:
```diff
@@ -359,5 +359,20 @@ public void TheOnTransitionEventFiresBeforeTheOnEntryEvent()
+
+            Assert.Throws(typeof(ArgumentException),  delegate { sm.Configure(State.A).SubstateOf(State.A); });
+        }
+        [Test]
```
Nit: missing newline.

HenningNT commented:
Thanks for the feedback!
I've fixed the nitpicks, and added a necessary test (which failed initially).
Also switched to Hashset instead of list, it reduced the line count

nblumhardt commented:
Nice! :-)

HenningNT commented:
Thanks!

Sendt fra Windows 10-telefonen min

Fra: Nicholas Blumhardt
Sendt: søndag 22. januar 2017 kl. 23.46
Til: dotnet-state-machine/stateless
Kopi: HenningNT; Author
Emne: Re: [dotnet-state-machine/stateless] Adding check for cyclicconfiguration (#125)

Nice! :-)
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.
